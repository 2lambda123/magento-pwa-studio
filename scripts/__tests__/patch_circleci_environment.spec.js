// This tests the script as a JS module (to achieve coverage analysis),
// and also as a separately running process (to mimic CI).
// These scripts aren't run through Babel, so neither are their tests;
// use require() here.
const { exec } = require('child_process');
const { promisify } = require('util');
const { resolve } = require('path');
const mockServer = require('./__helpers__/mock-http-server');

const patchCircleCIEnvironment = require('../patch_circleci_environment');

const execp = promisify(exec);

const patchScript = resolve(__dirname, '../patch_circleci_env');

const possiblePRVars = [
    'CIRCLE_PULL_REQUEST',
    'CIRCLE_PULL_REQUESTS',
    'CI_PULL_REQUEST',
    'CI_PULL_REQUESTS',
    'CIRCLE_PR_USERNAME',
    'CIRCLE_PR_NUMBER',
    'CIRCLE_PR_REPONAME'
];
const blankPRVars = possiblePRVars.reduce((out, name) => {
    out[name] = '';
    return out;
}, {});

/**
 * GitHub API simulation
 * In order to test in-process and out-of-process, we must expose a real
 * mock web server which a subprocess can contact, rather than mock with Jest.
 */
const GithubMock = (() => {
    const handler = jest.fn();
    const server = mockServer(handler);
    const api = {};
    api.reset = () => {
        handler.mockRestore();
        handler.mockReturnValue({
            status: 400,
            body: 'No mock response configured for this test run'
        });
    };
    api.start = async () => {
        api.url = await server.mount();
    };
    api.stop = async () => {
        delete api.url;
        return server.unmount();
    };
    api.responds = values => handler.mockReturnValue(values);
    api.getCalls = () => {
        return handler.mock.calls;
    };
    return api;
})();

beforeAll(GithubMock.start);
afterAll(GithubMock.stop);

beforeEach(GithubMock.reset);

/**
 * Test the script running in process (this provices code coverage)
 */
async function testInProcess(env) {
    const setEnv = jest.fn();
    const log = jest.fn();
    await patchCircleCIEnvironment({
        env,
        setEnv,
        log
    });
    return {
        stdout: setEnv.mock.calls
            .map(call => call.join('\n'))
            .join('\n')
            .trim(),
        stderr: log.mock.calls
            .map(call => call.join('\n'))
            .join('\n')
            .trim(),
        status: 0
    };
}

/**
 * Test the shell script which passes real OS objects, in a child process.
 * (this provides a stronger similarity to the way this works in CI)
 */
async function testOutOfProcess(env) {
    const { stdout, stderr, error } = await execp(patchScript, {
        encoding: 'utf8',
        env: Object.assign(
            {
                PATH: process.env.PATH
            },
            env
        )
    }).catch(e => e);
    return {
        stdout: stdout.trim(),
        stderr: stderr.trim(),
        status: (error && error.code) || 0
    };
}

async function testPatchWithEnv(env) {
    const mockedEnv = Object.assign(
        {
            GITHUB_GRAPHQL_ENDPOINT: GithubMock.url,
            DANGER_GITHUB_API_TOKEN: '##DANGER_GITHUB_API_TOKEN##'
        },
        env
    );
    // test both in-process functionality and child process functionality
    // (for code coverage and CI simulation)
    // return a big object for comparison and snapshotting
    return {
        inProcess: await testInProcess(mockedEnv),
        childProcess: await testOutOfProcess(mockedEnv)
    };
}

test('logs to stderr when graphql and/or github tokens are not in env', async () => {
    expect(
        await testPatchWithEnv({
            GITHUB_GRAPHQL_ENDPOINT: ''
        })
    ).toMatchSnapshot();
    expect(
        await testPatchWithEnv({
            DANGER_GITHUB_API_TOKEN: ''
        })
    ).toMatchSnapshot();
});

test('logs to stderr that PR env vars are already set and exits', async () => {
    function justOnePRVar(present) {
        return Object.assign(
            [
                'CIRCLE_PULL_REQUEST',
                'CIRCLE_PULL_REQUESTS',
                'CI_PULL_REQUEST',
                'CI_PULL_REQUESTS'
            ].reduce((out, name) => {
                out[name] =
                    name === present
                        ? 'https://github.com/someOwner/someRepo/pulls/4'
                        : '';
                return out;
            }, {}),
            {
                CIRCLE_PR_USERNAME: 'darkseid'
            }
        );
    }

    for (const envName of possiblePRVars) {
        const results = await testPatchWithEnv(justOnePRVar(envName));
        expect(results.childProcess.status).toBe(0);
        expect(results.inProcess).toEqual(results.childProcess);
        expect(results).toMatchSnapshot();
    }
});

test('logs to stderr some PR env vars exist but not all of them', async () => {
    GithubMock.responds({
        status: 200,
        body: {
            data: {
                repository: {
                    ref: {
                        associatedPullRequests: {
                            nodes: []
                        }
                    }
                }
            }
        }
    });
    const results = await testPatchWithEnv(
        Object.assign({}, blankPRVars, {
            CIRCLE_PROJECT_USERNAME: 'apokolips',
            CIRCLE_PROJECT_REPONAME: 'darkseid',
            CIRCLE_PR_USERNAME: 'new-genesis',
            CIRCLE_PR_REPONAME: 'orion',
            CIRCLE_BRANCH: 'nonexistent/branch'
        })
    );
    expect(results.childProcess.status).toBe(0);
    expect(results.inProcess).toEqual(results.childProcess);
    expect(results).toMatchSnapshot();
    expect(GithubMock.getCalls()).toMatchSnapshot();
});

test('logs to stderr if fetch network-errors', async () => {
    expect(
        await testPatchWithEnv({
            GITHUB_GRAPHQL_ENDPOINT: 'a bad url'
        })
    ).toMatchSnapshot();
    expect(GithubMock.getCalls()).toMatchSnapshot();
});

test('logs to stderr if response comes back non-200', async () => {
    GithubMock.responds({
        status: 200,
        body: {
            status: 404,
            body: {
                data: {
                    repository: {
                        ref: {
                            associatedPullRequests: {}
                        }
                    }
                }
            }
        }
    });
    const results = await testPatchWithEnv({
        CIRCLE_PROJECT_USERNAME: 'apokolips',
        CIRCLE_PROJECT_REPONAME: 'darkseid',
        CIRCLE_BRANCH: 'missing/branch'
    });
    expect(results.childProcess.status).toBe(0);
    expect(results.inProcess).toEqual(results.childProcess);
    expect(results).toMatchSnapshot();
    expect(GithubMock.getCalls()).toMatchSnapshot();
});

test('writes script that sets env vars to stdout if pr exists', async () => {
    GithubMock.responds({
        status: 200,
        body: {
            data: {
                repository: {
                    ref: {
                        associatedPullRequests: {
                            nodes: [
                                {
                                    url: 'https://github.com/found/me',
                                    number: 420
                                }
                            ]
                        }
                    }
                }
            }
        }
    });
    const results = await testPatchWithEnv({
        CIRCLE_PROJECT_USERNAME: 'apokolips',
        CIRCLE_PROJECT_REPONAME: 'darkseid',
        CIRCLE_BRANCH: 'missing/branch'
    });
    expect(results.childProcess.status).toBe(0);
    expect(results.inProcess).toEqual(results.childProcess);
    expect(results).toMatchSnapshot();
    expect(GithubMock.getCalls()).toMatchSnapshot();
});
