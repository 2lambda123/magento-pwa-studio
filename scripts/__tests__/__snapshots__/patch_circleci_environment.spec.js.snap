// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logs to stderr if fetch network-errors 1`] = `
Object {
  "childProcess": Object {
    "status": 0,
    "stderr": "Failed to contact GitHub API: TypeError: Only absolute URLs are supported",
    "stdout": "",
  },
  "inProcess": Object {
    "status": 0,
    "stderr": "Failed to contact GitHub API: TypeError: Only absolute URLs are supported",
    "stdout": "",
  },
}
`;

exports[`logs to stderr if fetch network-errors 2`] = `Array []`;

exports[`logs to stderr if response comes back non-200 1`] = `
Object {
  "childProcess": Object {
    "status": 0,
    "stderr": "Failed to find and export pull request variables using the GitHub
API. Check the environment setup step to see what env vars were provided.

Request: {
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"owner\\": \\"apokolips\\",
    \\"name\\": \\"darkseid\\",
    \\"ref\\": \\"refs/heads/missing/branch\\"
  }
}

Response: {\\"status\\":404,\\"body\\":{\\"data\\":{\\"repository\\":{\\"ref\\":{\\"associatedPullRequests\\":{}}}}}}

TypeError: Cannot read property 'repository' of undefined",
    "stdout": "",
  },
  "inProcess": Object {
    "status": 0,
    "stderr": "Failed to find and export pull request variables using the GitHub
API. Check the environment setup step to see what env vars were provided.

Request: {
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"owner\\": \\"apokolips\\",
    \\"name\\": \\"darkseid\\",
    \\"ref\\": \\"refs/heads/missing/branch\\"
  }
}

Response: {\\"status\\":404,\\"body\\":{\\"data\\":{\\"repository\\":{\\"ref\\":{\\"associatedPullRequests\\":{}}}}}}

TypeError: Cannot read property 'repository' of undefined",
    "stdout": "",
  },
}
`;

exports[`logs to stderr if response comes back non-200 2`] = `
Array [
  Array [
    Object {
      "body": "{
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"owner\\": \\"apokolips\\",
    \\"name\\": \\"darkseid\\",
    \\"ref\\": \\"refs/heads/missing/branch\\"
  }
}",
      "method": "POST",
    },
  ],
  Array [
    Object {
      "body": "{
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"owner\\": \\"apokolips\\",
    \\"name\\": \\"darkseid\\",
    \\"ref\\": \\"refs/heads/missing/branch\\"
  }
}",
      "method": "POST",
    },
  ],
]
`;

exports[`logs to stderr some PR env vars exist but not all of them 1`] = `
Object {
  "childProcess": Object {
    "status": 0,
    "stderr": "Some pull request environment variables are present, and others
are not. DangerCI may not correctly identify this as a pull request. Check the
CircleCI \\"Spin up environment\\" step to see build environment variables.

        Pull request from: new-genesis/orion
        instead of:        apokolips/darkseid


Found no pull requests on GitHub for nonexistent/branch.",
    "stdout": "",
  },
  "inProcess": Object {
    "status": 0,
    "stderr": "Some pull request environment variables are present, and others
are not. DangerCI may not correctly identify this as a pull request. Check the
CircleCI \\"Spin up environment\\" step to see build environment variables.

        Pull request from: new-genesis/orion
        instead of:        apokolips/darkseid


Found no pull requests on GitHub for nonexistent/branch.",
    "stdout": "",
  },
}
`;

exports[`logs to stderr some PR env vars exist but not all of them 2`] = `
Array [
  Array [
    Object {
      "body": "{
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"owner\\": \\"apokolips\\",
    \\"name\\": \\"darkseid\\",
    \\"ref\\": \\"refs/heads/nonexistent/branch\\"
  }
}",
      "method": "POST",
    },
  ],
  Array [
    Object {
      "body": "{
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"owner\\": \\"apokolips\\",
    \\"name\\": \\"darkseid\\",
    \\"ref\\": \\"refs/heads/nonexistent/branch\\"
  }
}",
      "method": "POST",
    },
  ],
]
`;

exports[`logs to stderr that PR env vars are already set and exits 1`] = `
Object {
  "childProcess": Object {
    "status": 0,
    "stderr": "No change: CircleCI has already populated pull request variables
for this build, so 'danger ci' should run as expected.

Pull Request: https://github.com/someOwner/someRepo/pulls/4 from darkseid",
    "stdout": "",
  },
  "inProcess": Object {
    "status": 0,
    "stderr": "No change: CircleCI has already populated pull request variables
for this build, so 'danger ci' should run as expected.

Pull Request: https://github.com/someOwner/someRepo/pulls/4 from darkseid",
    "stdout": "",
  },
}
`;

exports[`logs to stderr that PR env vars are already set and exits 2`] = `
Object {
  "childProcess": Object {
    "status": 0,
    "stderr": "No change: CircleCI has already populated pull request variables
for this build, so 'danger ci' should run as expected.

Pull Request: https://github.com/someOwner/someRepo/pulls/4 from darkseid",
    "stdout": "",
  },
  "inProcess": Object {
    "status": 0,
    "stderr": "No change: CircleCI has already populated pull request variables
for this build, so 'danger ci' should run as expected.

Pull Request: https://github.com/someOwner/someRepo/pulls/4 from darkseid",
    "stdout": "",
  },
}
`;

exports[`logs to stderr that PR env vars are already set and exits 3`] = `
Object {
  "childProcess": Object {
    "status": 0,
    "stderr": "No change: CircleCI has already populated pull request variables
for this build, so 'danger ci' should run as expected.

Pull Request: https://github.com/someOwner/someRepo/pulls/4 from darkseid",
    "stdout": "",
  },
  "inProcess": Object {
    "status": 0,
    "stderr": "No change: CircleCI has already populated pull request variables
for this build, so 'danger ci' should run as expected.

Pull Request: https://github.com/someOwner/someRepo/pulls/4 from darkseid",
    "stdout": "",
  },
}
`;

exports[`logs to stderr that PR env vars are already set and exits 4`] = `
Object {
  "childProcess": Object {
    "status": 0,
    "stderr": "No change: CircleCI has already populated pull request variables
for this build, so 'danger ci' should run as expected.

Pull Request: https://github.com/someOwner/someRepo/pulls/4 from darkseid",
    "stdout": "",
  },
  "inProcess": Object {
    "status": 0,
    "stderr": "No change: CircleCI has already populated pull request variables
for this build, so 'danger ci' should run as expected.

Pull Request: https://github.com/someOwner/someRepo/pulls/4 from darkseid",
    "stdout": "",
  },
}
`;

exports[`logs to stderr that PR env vars are already set and exits 5`] = `
Object {
  "childProcess": Object {
    "status": 0,
    "stderr": "Some pull request environment variables are present, and others
are not. DangerCI may not correctly identify this as a pull request. Check the
CircleCI \\"Spin up environment\\" step to see build environment variables.

        Pull request from: darkseid/undefined
        instead of:        undefined/undefined


Failed to find and export pull request variables using the GitHub
API. Check the environment setup step to see what env vars were provided.

Request: {
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"ref\\": \\"refs/heads/undefined\\"
  }
}

Response: No mock response configured for this test run

Error: HTTP 400",
    "stdout": "",
  },
  "inProcess": Object {
    "status": 0,
    "stderr": "Some pull request environment variables are present, and others
are not. DangerCI may not correctly identify this as a pull request. Check the
CircleCI \\"Spin up environment\\" step to see build environment variables.

        Pull request from: darkseid/undefined
        instead of:        undefined/undefined


Failed to find and export pull request variables using the GitHub
API. Check the environment setup step to see what env vars were provided.

Request: {
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"ref\\": \\"refs/heads/undefined\\"
  }
}

Response: No mock response configured for this test run

Error: HTTP 400",
    "stdout": "",
  },
}
`;

exports[`logs to stderr that PR env vars are already set and exits 6`] = `
Object {
  "childProcess": Object {
    "status": 0,
    "stderr": "Some pull request environment variables are present, and others
are not. DangerCI may not correctly identify this as a pull request. Check the
CircleCI \\"Spin up environment\\" step to see build environment variables.

        Pull request from: darkseid/undefined
        instead of:        undefined/undefined


Failed to find and export pull request variables using the GitHub
API. Check the environment setup step to see what env vars were provided.

Request: {
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"ref\\": \\"refs/heads/undefined\\"
  }
}

Response: No mock response configured for this test run

Error: HTTP 400",
    "stdout": "",
  },
  "inProcess": Object {
    "status": 0,
    "stderr": "Some pull request environment variables are present, and others
are not. DangerCI may not correctly identify this as a pull request. Check the
CircleCI \\"Spin up environment\\" step to see build environment variables.

        Pull request from: darkseid/undefined
        instead of:        undefined/undefined


Failed to find and export pull request variables using the GitHub
API. Check the environment setup step to see what env vars were provided.

Request: {
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"ref\\": \\"refs/heads/undefined\\"
  }
}

Response: No mock response configured for this test run

Error: HTTP 400",
    "stdout": "",
  },
}
`;

exports[`logs to stderr that PR env vars are already set and exits 7`] = `
Object {
  "childProcess": Object {
    "status": 0,
    "stderr": "Some pull request environment variables are present, and others
are not. DangerCI may not correctly identify this as a pull request. Check the
CircleCI \\"Spin up environment\\" step to see build environment variables.

        Pull request from: darkseid/undefined
        instead of:        undefined/undefined


Failed to find and export pull request variables using the GitHub
API. Check the environment setup step to see what env vars were provided.

Request: {
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"ref\\": \\"refs/heads/undefined\\"
  }
}

Response: No mock response configured for this test run

Error: HTTP 400",
    "stdout": "",
  },
  "inProcess": Object {
    "status": 0,
    "stderr": "Some pull request environment variables are present, and others
are not. DangerCI may not correctly identify this as a pull request. Check the
CircleCI \\"Spin up environment\\" step to see build environment variables.

        Pull request from: darkseid/undefined
        instead of:        undefined/undefined


Failed to find and export pull request variables using the GitHub
API. Check the environment setup step to see what env vars were provided.

Request: {
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"ref\\": \\"refs/heads/undefined\\"
  }
}

Response: No mock response configured for this test run

Error: HTTP 400",
    "stdout": "",
  },
}
`;

exports[`logs to stderr when graphql and/or github tokens are not in env 1`] = `
Object {
  "childProcess": Object {
    "status": 0,
    "stderr": "No GITHUB_GRAPHQL_ENDPOINT present in environment.
Cannot call GitHub to populate missing PR variables.
Failed to contact GitHub API: TypeError: Only absolute URLs are supported",
    "stdout": "",
  },
  "inProcess": Object {
    "status": 0,
    "stderr": "No GITHUB_GRAPHQL_ENDPOINT present in environment.
Cannot call GitHub to populate missing PR variables.
Failed to contact GitHub API: TypeError: Only absolute URLs are supported",
    "stdout": "",
  },
}
`;

exports[`logs to stderr when graphql and/or github tokens are not in env 2`] = `
Object {
  "childProcess": Object {
    "status": 0,
    "stderr": "No DANGER_GITHUB_API_TOKEN present in environment.
Cannot call GItHub to populate missing PR variables.
Failed to find and export pull request variables using the GitHub
API. Check the environment setup step to see what env vars were provided.

Request: {
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"ref\\": \\"refs/heads/undefined\\"
  }
}

Response: No mock response configured for this test run

Error: HTTP 400",
    "stdout": "",
  },
  "inProcess": Object {
    "status": 0,
    "stderr": "No DANGER_GITHUB_API_TOKEN present in environment.
Cannot call GItHub to populate missing PR variables.
Failed to find and export pull request variables using the GitHub
API. Check the environment setup step to see what env vars were provided.

Request: {
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"ref\\": \\"refs/heads/undefined\\"
  }
}

Response: No mock response configured for this test run

Error: HTTP 400",
    "stdout": "",
  },
}
`;

exports[`writes script that sets env vars to stdout if pr exists 1`] = `
Object {
  "childProcess": Object {
    "status": 0,
    "stderr": "Acquired missing pull request data from GitHub. This is PR #420.
 URL: https://github.com/found/me. Added environment variables:

CI_PULL_REQUEST=\\"https://github.com/found/me\\"
CI_PULL_REQUESTS=\\"https://github.com/found/me\\"
CIRCLE_PULL_REQUEST=\\"https://github.com/found/me\\"
CIRCLE_PULL_REQUESTS=\\"https://github.com/found/me\\"
CIRCLE_PR_NUMBER=\\"420\\"",
    "stdout": "CI_PULL_REQUEST=\\"https://github.com/found/me\\"
CI_PULL_REQUESTS=\\"https://github.com/found/me\\"
CIRCLE_PULL_REQUEST=\\"https://github.com/found/me\\"
CIRCLE_PULL_REQUESTS=\\"https://github.com/found/me\\"
CIRCLE_PR_NUMBER=\\"420\\"",
  },
  "inProcess": Object {
    "status": 0,
    "stderr": "Acquired missing pull request data from GitHub. This is PR #420.
 URL: https://github.com/found/me. Added environment variables:

CI_PULL_REQUEST=\\"https://github.com/found/me\\"
CI_PULL_REQUESTS=\\"https://github.com/found/me\\"
CIRCLE_PULL_REQUEST=\\"https://github.com/found/me\\"
CIRCLE_PULL_REQUESTS=\\"https://github.com/found/me\\"
CIRCLE_PR_NUMBER=\\"420\\"",
    "stdout": "CI_PULL_REQUEST=\\"https://github.com/found/me\\"
CI_PULL_REQUESTS=\\"https://github.com/found/me\\"
CIRCLE_PULL_REQUEST=\\"https://github.com/found/me\\"
CIRCLE_PULL_REQUESTS=\\"https://github.com/found/me\\"
CIRCLE_PR_NUMBER=\\"420\\"",
  },
}
`;

exports[`writes script that sets env vars to stdout if pr exists 2`] = `
Array [
  Array [
    Object {
      "body": "{
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"owner\\": \\"apokolips\\",
    \\"name\\": \\"darkseid\\",
    \\"ref\\": \\"refs/heads/missing/branch\\"
  }
}",
      "method": "POST",
    },
  ],
  Array [
    Object {
      "body": "{
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"owner\\": \\"apokolips\\",
    \\"name\\": \\"darkseid\\",
    \\"ref\\": \\"refs/heads/missing/branch\\"
  }
}",
      "method": "POST",
    },
  ],
]
`;
